nustate/operation/open_ports.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package operation
+
+import (
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/nustate/model"
+	"github.com/juju/juju/nustate/operation/precondition"
+	"github.com/juju/juju/nustate/persistence"
+	"github.com/juju/juju/nustate/persistence/transaction"
+)
+
+// Changes is implemented by types that can provide a set of transaction
+// elements for applying a complex, multi-model mutation operation.
+type Changes interface {
+	transaction.Element
+	Changes(transaction.Context) (transaction.ModelTxn, error)
+}
+
+// We could use the concrete type here but perhaps an interface would make
+// mocking easier at the facade level?
+type MutateMachinePortRangesOperation interface {
+	Changes
+
+	ForUnit(string) MutateUnitPortRangesOperation
+}
+
+type mutateMachinePortRangesOp struct {
+	transaction.BuildingBlock
+
+	store     persistence.Store
```

Comment:
I think this shouldn't be `persistence.Store` but `Store`, this should be a representation of what you want, not what the persistence gives you. That would also solve the `persistence.Store` mega interface.

Considering that the underlying store would implement that interface (or compile error) then we should be ok with tight smaller interfaces.

Reply:
I agree. Ideally, we would have many small interfaces in the `operation` package but given that we will probably have hundreds of operations in there, having one interface per operation (so we can apply the interface segregation principle) might be overkill (both in terms of maintainance as well as binary size as the go compiler would potentially generate a large itab section). 

This is why I suggested the idea of defining store interfaces per entity type and compose them into larger interfaces where required. If we limit the things we need the store to do (it's should be pretty much be `lookXByY` methods, e.g. `FindMachineByID`, `FindApplicationByName` etc.) we should end up with really small interfaces which might get us closer to what you propose.